---
title: "Loading Gene Clusters From Generic Feature Format (GFF) Files"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Loading Gene Clusters From Generic Feature Format (GFF) Files}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
library(knitr)
library(geneviewer)
library(dplyr)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

### Intro

This tutorial demonstrates creating gene cluster visualizations from Generic Feature Format (GFF) files using `geneviewer`. You can obtain .gff files from a variety of platforms such as [NCBI](https://www.ncbi.nlm.nih.gov/genbank/), [Ensembl](https://www.ensembl.org/) or [UCSC](https://genome.ucsc.edu/). This tutorial demonstrates how to load and visualize several viral genomes in .gff format using geneviewer. 

### Materials 

The .gff files were retrieved from the [Ensembl](https://www.ensembl.org/) website by searching with their respective GenBank identifiers and downloading the records in GFF3 format. Alternatively, the files can also be directly accessed from the [geneviewer-tutorials](https://github.com/nvelden/geneviewer-tutorials/tree/main/Loading%20GFF%20files) repository.

### Loading GFF files

We can load the .gff files into R using the read_gff function from the geneviewer package. To do this, we can either load each file individually by specifying its file path, or load all files at once by specifying the directory that contains all the .gff files.

In the example below, we load all .gff files from a specified directory using the `fields` parameter to select specific fields for loading. If no fields are specified, all fields will be loaded by default. Using the `dplyr` package, we filter the data to select entries where the 'type' column contains 'CDS'. In addition we add an extra column which maps each GenBank ID in the `filename` column to its corresponding viral name.

```{r eval=FALSE, results='hide', warning=FALSE}
library(geneviewer)
library(dplyr)
# change the path to where you have saved the 
# file or the directory containing all .gff files
folder_path <- "~/path/to/folder/"
gff <- read_gff(
  folder_path, 
  fields = c("source", "type", "start", "end", "strand", "Name")
  ) %>%
  dplyr::filter(type == "CDS")

# Add viral names
virus_names <- c(
  GU071086 = "Marseillevirus", 
  HQ113105.1 = "Lausannevirus",
  KF261120 = "Cannesvirus",
  KF483846 = "Tunisvirus"
)
gff <- gff %>% mutate(Name = virus_names[filename]) 

View(gff) # Inspect the data frame in Rstudio
```

```{r echo = FALSE, results='asis', warning=FALSE}
library(geneviewer)
library(dplyr)
# change the path to where you have saved the file
folder_path <- "~/Documents/2023 github/geneviewer/gff_files/marseilleviridae/"
gff <- read_gff(folder_path, fields = c("source", "type", "start", "end", "strand", "Name")) %>%
  dplyr::filter(type == "CDS")
virus_names <- c(
  GU071086 = "Marseillevirus", 
  HQ113105.1 = "Lausannevirus",
  KF261120 = "Cannesvirus",
  KF483846 = "Tunisvirus"
)
gff <- gff %>% dplyr::mutate(Virus = virus_names[filename]) 
kable(head(gff))
```

We can now visualize the genomic data using `geneviewer`. The chart displays the start and end positions of genes for each viral genome oriented according to strand location. A custom title is added to the chart, and the gene clusters are labeled with their viral names. The axis is set to `range` to facilitate size comparisons among the genomes. Additionally, the visual representation of the genes is simplified by removing outlines (strokes) and using smaller markers. As a final touch we change the tooltip to only display the gene Name. 

```{r echo = TRUE, results='asis', warning=FALSE}
GC_chart(gff, 
         start = "start", 
         end = "end", 
         strand = "strand", 
         cluster = "Virus",
         height = "400px",
         ) %>%
  GC_title(
    "<i>Marseilleviridae</i> viral genomes", 
    height = "40px") %>%
  GC_clusterLabel() %>%
  GC_scale(axis_type = "range") %>%
  GC_genes(
    group = "filename", 
    stroke = "none", 
    marker_size = "small"
    ) %>%
  GC_tooltip(formatter = "<b>Gene:</b> {Name}")
```
